RecoCTPPS/PixelLocal/plugins/CTPPSPixelLocalTrackProducer.cc

```diff
@@ -0,0 +1,289 @@
+    CTPPSPixelDetId tmpRomanPotId = CTPPSPixelDetId(recHitSet.detId()).getRPId();
+    uint32_t hitOnPlane = recHitSet.size();
+
+    //Get the number of hits per pot
+    if(mapHitPerPot.find(tmpRomanPotId)==mapHitPerPot.end()){
+      mapHitPerPot[tmpRomanPotId] = hitOnPlane;
+    }
+    else mapHitPerPot[tmpRomanPotId]+=hitOnPlane;
+
+    //check is the plane occupancy is too high and save the corresponding pot
+    if(maxHitPerPlane_>=0 && hitOnPlane>(uint32_t)maxHitPerPlane_){
+      if(verbosity_>2) edm::LogInfo("CTPPSPixelLocalTrackProducer")<<" ---> To many hits in the plane, pot will be excluded from tracking cleared";
+      listOfPotWithHighOccupancyPlanes.push_back(tmpRomanPotId);
+    }
+  }
+
+  //remove rechit for pot with too many hits or containing planes with too many hits
+  for(const auto & recHitSet : recHitVector){
+    const auto tmpDetectorId = CTPPSPixelDetId(recHitSet.detId());
+    const auto tmpRomanPotId = tmpDetectorId.getRPId();
+
+    if((maxHitPerRomanPot_>=0 && mapHitPerPot[tmpRomanPotId]>(uint32_t)maxHitPerRomanPot_) || 
+      find (listOfPotWithHighOccupancyPlanes.begin(), listOfPotWithHighOccupancyPlanes.end(), tmpRomanPotId) != listOfPotWithHighOccupancyPlanes.end() ){
+      edm::DetSet<CTPPSPixelRecHit> &tmpDetSet = recHitVector[tmpDetectorId];
+      tmpDetSet.clear();
+    }
+
+  }
+
+  edm::DetSetVector<CTPPSPixelLocalTrack> foundTracks;
+
+  // Pattern finder
+
+  patternFinder_->clear();
+  patternFinder_->setHits(&recHitVector);
+  patternFinder_->setGeometry(&geometry);
+  patternFinder_->findPattern();
+  std::vector<RPixDetPatternFinder::Road> patternVector = patternFinder_->getPatterns();
+
+  //loop on all the patterns
+  int numberOfTracks = 0;
+
+  for(const auto & pattern : patternVector){
+    CTPPSPixelDetId firstHitDetId = CTPPSPixelDetId(pattern.at(0).detId);
+    CTPPSPixelDetId romanPotId = firstHitDetId.getRPId();
+    
+    std::map<CTPPSPixelDetId, std::vector<RPixDetPatternFinder::PointInPlane> > hitOnPlaneMap; //hit of the pattern organized by plane
+
+    //loop on all the hits of the pattern
+    for(const auto & hit : pattern){
+      CTPPSPixelDetId hitDetId = CTPPSPixelDetId(hit.detId);
+      CTPPSPixelDetId tmpRomanPotId = hitDetId.getRPId();
+      
+
+      if(tmpRomanPotId!=romanPotId){ //check that the hits belong to the same tracking station
+        throw cms::Exception("CTPPSPixelLocalTrackProducer") << "Hits in the pattern must belong to the same tracking station";
+      }
+
+      if(hitOnPlaneMap.find(hitDetId)==hitOnPlaneMap.end()){ //add the plane key in case it is the first hit of that plane
+        std::vector<RPixDetPatternFinder::PointInPlane> hitOnPlane;
+        hitOnPlane.push_back(hit);
+        hitOnPlaneMap[hitDetId] = hitOnPlane;
+      }
+      else hitOnPlaneMap[hitDetId].push_back(hit); //add the hit to an existing plane key
+    
+    }
+
+    trackFinder_->clear();
```

Comment:
could this clear be moved to the end of the loop as to clean up the memory when the module is done running an event?

Reply:
Unfortunately, `std::vector<>::clear` does not release the memory held by the container. It does delete the objects, of course, and if those objects have pointers to items on the heap then that memory would be recovered. To recover the memory, the `clear` function would have to do
```
localTrackVector_ = std::vector<CTPPSPixelLocalTrack>();
```